### PR TITLE
Add rocBLAS fp16 GEMM take2

### PR DESCRIPTION
**Pull request by whchung**
Reincarnation the effort to add rocBLAS fp16 GEMM into ROCm StreamExecutor implementation.

### CODE REVIEW

**Comment by whchung** [comment]
cc @amcamd for awareness.

**Comment by whchung** [comment]
interesting... on my box it builds fine but we meet:

```
/usr/lib/gcc/x86_64-linux-gnu/5/include/f16cintrin.h:59:1: error: inlining failed in call to always_inline 'short unsigned int _cvtss_sh(float, int)': target specific option mismatch
 _cvtss_sh (float __F, const int __I)
 ^
tensorflow/stream_executor/rocm/rocm_blas.cc:1473:44: error: called from here
   const rocblas_half alpha_half = _cvtss_sh(alpha, 0);
```

on CI node. looking into it.

**Comment by deven-amd** [comment]
This is the Eigen routine for converting float to half

https://github.com/eigenteam/eigen-git-mirror/blob/master/Eigen/src/Core/arch/GPU/Half.h#L407

It gets called when you instantiate Eigen::half with a float....interestingly it calls the same `__cvtss_sh` but seems to work okay.

**Comment by whchung** [comment]
it's kind of weird. with the PR in its current shape I've already moved forward quite a bit and looking into other issues in `MIOpen`... will have to figure this out anyhow

File: tensorflow/stream_executor/rocm/rocm_blas.cc [hunk 1]
```diff
@@ -1438,8 +1442,42 @@ bool ROCMBlas::DoBlasGemm(
...
+      LOG(WARNING) << "GEMM ldb (" << ldb << ") was smaller than k (" << k
+                   << ") (no transpose case); precondition violation";
+    }
+  } else {
+    if (ldb < static_cast<int64>(n)) {
+      LOG(WARNING) << "GEMM ldb was smaller than n (transpose case); "
+                      "precondition violation";
+    }
+  }
+  const rocblas_half alpha_half = _cvtss_sh(alpha, 0);
```
Comment (deven-amd) [inline tensorflow/stream_executor/rocm/rocm_blas.cc]: How about going the `alpha_half = Eigen::half(alpha);` routine instead? 

That would keep things consistent with the other parameters (`A`, `B`, `C`)...that way all input params need to have the same `Eigen::half` to `rocblas_half` conversion (currently just a static cast), and if there is any memory mapping we need to apply when converting `Eigen::half` to `rocblas_half`, we can apply it to all input params.

Comment (whchung) [reply tensorflow/stream_executor/rocm/rocm_blas.cc]: thanks. let me give it a shot.

Comment (whchung) [reply tensorflow/stream_executor/rocm/rocm_blas.cc]: `Eigen::half` has conversion operators to `unsigned short` i assume?

Comment (deven-amd) [reply tensorflow/stream_executor/rocm/rocm_blas.cc]: Yes....the `.x` member is the unsigned short.

File: tensorflow/stream_executor/rocm/rocm_blas.cc [hunk 2]
```diff
@@ -1438,8 +1442,42 @@ bool ROCMBlas::DoBlasGemm(
...
+  const rocblas_half beta_half = _cvtss_sh(beta, 0);
+  return DoBlasInternal(
+      wrap::rocblas_hgemm, stream, true /* = pointer_mode_host */,
+      ROCMBlasTranspose(transa), ROCMBlasTranspose(transb), m, n, k,
+      &alpha_half, static_cast<const rocblas_half*>(&ROCMMemory(a)->x), lda,
```
Comment (deven-amd) [inline tensorflow/stream_executor/rocm/rocm_blas.cc]: do we need to have the `->x` in `&ROCMMemory(a)->x`?

I would expect both `&ROCMMemory(a)->x` and `ROCMMemory(a)`  to point to the same address (`x` being the only data member within `Eigen::half`)

Comment (whchung) [reply tensorflow/stream_executor/rocm/rocm_blas.cc]: IIRC that was the first implementation and g++ 5.4.x threw a lot of stuffs to my face J

I can do another try tho.

**Comment by whchung** [comment]
@deven-amd indeed g++ 5.4.0 doesn't like your proposal:

```
tensorflow/stream_executor/rocm/rocm_blas.cc:1479:53: error: invalid static_cast from type 'const Eigen::half*' to type 'const rocblas_half* {aka const short unsigned int*}'
       static_cast<const rocblas_half*>(ROCMMemory(a)), lda,
```

**Pull request merged**